_drafts/zero-to-main-libc.md

```diff
@@ -0,0 +1,385 @@
+
+```c
+int close(int fd) {
+  return -1;
+}
+```
+
+#### write
+
+This is where things get interesting! `write` writes up to `count` bytes from the buffer starting at `buf` to the file referred to by the file descriptor `fd`.
+
+Functions like `printf` rely on `write` to write bytes to `STDOUT`. In our case, we will want those bytes to be written to serial instead.
+
+On the SAMD21 chip we are using, writing bytes to serial is done using the `usart_serial_putchar` function. We can use it to implement `write`:
+
+```c
+static struct usart_module stdio_uart_module;
+
+int _write (int fd, char *buf, int count) {
+  int written = 0;
+
+  for (; count != 0; --count) {
+    if (usart_serial_putchar(&stdio_uart_module, (uint8_t)*buf++)) {
+      return -1;
+    }
+    ++written;
+  }
+  return written;
+}
+```
+
+We’ll also need to initialize the USART peripheral prior to calling printf:
+
+```c
+static void serial_init(void) {
+  struct usart_config usart_conf;
+
+  usart_get_config_defaults(&usart_conf);
+  usart_conf.mux_setting = USART_RX_3_TX_2_XCK_3;
+  usart_conf.pinmux_pad0 = PINMUX_UNUSED;
+  usart_conf.pinmux_pad1 = PINMUX_UNUSED;
+  usart_conf.pinmux_pad2 = PINMUX_PB22D_SERCOM5_PAD2;
+  usart_conf.pinmux_pad3 = PINMUX_PB23D_SERCOM5_PAD3;
+
+  usart_serial_init(&stdio_uart_module, SERCOM5, &usart_conf);
+  usart_enable(&stdio_uart_module);
+}
+
+int main() {
+  serial_init();
+
+  printf("Hello, World!\n");
+  while (1) {}
+}
+```
+
+#### read
+
+`read`  attempts to read up to `count` bytes from file descriptor `fd` into the buffer at `buf`. 
+
+Similarly to `write`, we want `read` to read bytes from serial:
+
+```c
+int _read (int fd, char *buf, int count) {
+  int read = 0;
+
+  for (; count > 0; --count) {
+    usart_serial_getchar(&stdio_uart_module, (uint8_t *)buf++);
+    read++;
+  }
+
+  return read;
+}
+```
+
+#### sbrk
+
+`sbrk` increases the program’s data space by `increment` bytes. In other words, it increases the size of the heap.
+
+What does `printf` have to do with the heap, you will justly ask? It turns out that newlib’s `printf` implementations allocates data on the heap and depends on a working `malloc` implementation. 
+
+The source for `printf` is hard to follow, but you will find that indeed [it calls malloc](https://github.com/bminor/newlib/blob/6497fdfaf41d47e835fdefc78ecb0a934875d7cf/newlib/libc/stdio/vfprintf.c#L226)!
+
+Here’s a simple implementation of `sbrk`:
+
+```
+caddr_t _sbrk(int incr) {
+  static unsigned char *heap = HEAP_START;
+  unsigned char *prev_heap = heap;
+  heap += incr;
+  return (caddr_t) prev_heap;
+}
+```
+
+More often than not, we want the heap to use all the RAM not used by anything else. We therefore set `HEAP_START` to the first address not spoken for in our linker script. In our [previous post](https://interrupt.memfault.com/blog/how-to-write-linker-scripts-for-firmware#complete-linker-script), we had added the `_end` variable in our linker script to that end.
```

Comment:
OT: What I don't like about this approach is as you add more data to your code, the heap silently shrinks on you increasing the chances of OOMs. Personally a fan of using a hardcoded static buffer in the C code itself

Reply:
You have a point. That would also be simpler w.r.t. the linker script (no need for the variable). At the same time, this is good enough and I'd rather not rewrite it.